modules/025-static-routing-manager/images/agent/pkg/controller/iprules_reconciler_agent.go

```diff
@@ -0,0 +1,818 @@
+	log.Debug(fmt.Sprintf("[NIRSReconciler] Getting all IPRules from our node"))
+	actualIPRulesOnNode, err = getActualIPRuleEntryMapFromNode()
+	if err != nil {
+		log.Error(err, fmt.Sprintf("[NIRSReconciler] unable to get Actual IPRules from node"))
+		return true, err
+	}
+	if len(actualIPRulesOnNode) == 0 {
+		log.Debug(fmt.Sprintf("[NIRSReconciler] There are no IPRules with Realm=" + strconv.Itoa(v1alpha1.D8Realm)))
+	}
+
+	for _, nirs := range nirsList.Items {
+		nirsSummary := nirsSummaryInit()
+		// Gathering facts
+		log.Debug(fmt.Sprintf("[NIRSReconciler] Starting gather facts about nirs %v", nirs.Name))
+		if nirsSummary.discoverFacts(nirs, &globalDesiredIPRulesForNode, &actualIPRulesOnNode, log) {
+			(*nirsMap)[nirs.Name] = nirsSummary
+			continue
+		}
+
+		// Actions: add IPRules
+		if len(nirsSummary.desiredIPRulesToAddByNIRS) > 0 {
+			log.Debug(fmt.Sprintf("[NIRSReconciler] Starting adding IPRules to the node"))
+			nirsSummary.addIPRules(&actualIPRulesOnNode, log)
+		}
+
+		(*nirsMap)[nirs.Name] = nirsSummary
+	}
+
+	// Actions: delete IPRules and finalizers
+	nirsMap.deleteIPRulesAndFinalizers(globalDesiredIPRulesForNode, actualIPRulesOnNode, log)
+
+	// Actions: Deleting orphan IPRules (with realm 216) that are not mentioned in any NIRS
+	deleteOrphanIPRules(globalDesiredIPRulesForNode, actualIPRulesOnNode, log)
+
+	// Generate new condition for each processed nirs
+	log.Debug(fmt.Sprintf("[NIRSReconciler] Starting generate new conditions"))
+	shouldRequeue := nirsMap.generateNewCondition()
+
+	// Update state in k8s for each processed nirs
+	log.Debug(fmt.Sprintf("[NIRSReconciler] Starting updating resourses in k8s"))
+	nirsMap.updateStateInK8S(ctx, cl, log)
+
+	return shouldRequeue, nil
+}
+
+func periodicalRunEventIPRuleReconcile(
+	ctx context.Context,
+	cfg config.Options,
+	cl client.Client,
+	log logger.Logger,
+	nodeName string,
+) {
+	ticker := time.NewTicker(cfg.PeriodicReconciliationInterval * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			log.Debug(fmt.Sprintf("[NIRSReconciler] Starts a periodic reconciliation (initiated by a timer)"))
+			_, err := runEventIPRuleReconcile(ctx, cl, log, nodeName)
+			if err != nil {
+				log.Error(err, fmt.Sprintf("[NIRSReconciler] an error occurred while IPRule reconcile"))
+			}
+			log.Debug(fmt.Sprintf("[NIRSReconciler] Ends a periodic reconciliation (initiated by a timer)"))
+		case <-ctx.Done():
+			log.Debug(fmt.Sprintf("[NIRSReconciler] Completion of periodic reconciliations"))
+			return
+		}
+	}
+}
+
+// IPRuleEntry: type, service functions and methods
+
+type IPRuleEntry struct {
+	Priority   int
+	Invert     bool
+	Src        string
+	Dst        string
+	IPProto    int
+	SPortRange *v1alpha1.PortRange
+	DPortRange *v1alpha1.PortRange
+	Tos        string
+	FWMark     string
+	IifName    string
+	OifName    string
+	UIDRange   *v1alpha1.UIDRange
+	Table      int
+}
+
+func (ire *IPRuleEntry) getHash() string {
```

Comment:
https://github.com/mitchellh/hashstructure

Reply:
Our hash func is convenient for debug, have merci :).

Reply:
We already use hashstructure in other projects. I believe it is safe.